src/core/Val.ml

```diff
@@ -1246,44 +1246,56 @@ struct
       let _, s' = Dir.unleash fhcom.dir in
 
       (* This is C_M in [F], with an extra parameter `phi` to get along with NbE. *)
-      let cap_aux phi el = make_cap (Dir.act phi fhcom.dir) (Value.act phi fhcom.cap) (CompSys.act phi fhcom.sys) el in
+      let cap_aux phi el =
+        let mdir = Dir.act phi fhcom.dir in
+        let ty = Value.act phi fhcom.cap in
+        let msys = CompSys.act phi fhcom.sys in
+        make_cap mdir ty msys el
+      in
 
       (* This serves as `O` and the diagonal face in [F]
        * for the coherence conditions in `fhcom.sys` and `s=s'`. *)
-      let hcom_template phi y_dest ty = make_hcom
-          (Dir.make (I.act phi r) y_dest) ty
-          (Value.act phi fhcom.cap) (CompSys.act phi fhcom.sys)
+      let hcom_template phi y_dest ~ty =
+        make_hcom
+          (Dir.make (I.act phi r) y_dest)
+          ty
+          (Value.act phi cap)
```

Comment:
Here was the extremely subtle bug: I believe that `hcom_template` should be using the outer `hcom`'s cap and system, not the inner `fcom`'s cap and system.

Sorry for hiding this change amidst formatting adjustments; I made these adjustments while I was "Getting Down" with the code and trying to understand it.

I'm still not completely confident that this matches the algorithm in @cangiuli's summary of @mortberg's `hcom-in-fcom` note (and I don't have the Immortal @favonia Note on hand). But I think this is closer.

Reply:
Thanks!